### PR TITLE
optimize area for reservation station by splitting instructions

### PR DESCRIPTION
**Pull request by RCD-Y**

### CODE REVIEW

File: src/main/scala/gemmini/ReservationStation.scala [hunk 1]
```diff
@@ -404,33 +374,50 @@ class ReservationStation[T <: Data : Arithmetic, U <: Data, V <: Data](config: G
   }
 
   // Issue commands which are ready to be issued
-  Seq((ldq, io.issue.ld), (stq, io.issue.st), (exq, io.issue.ex)).foreach { case (q, io) =>
-    val issue_valids = entries.map(e => e.valid && e.bits.ready() && !e.bits.issued && e.bits.q === q)
+  Seq((ldq, io.issue.ld, entries_ld), (stq, io.issue.st, entries_st), (exq, io.issue.ex, entries_ex))
```
Comment (RCD-Y) [inline src/main/scala/gemmini/ReservationStation.scala:377]: note to self: should've used the previous version of this, since the new ID scheme wasn't applied to elsewhere. if something failed, it's here

File: src/main/scala/gemmini/Configs.scala [hunk 2]
```diff
@@ -51,8 +51,7 @@ object GemminiConfigs {
     has_nonlinear_activations = true,
 
     // Reservation station entries
-    reservation_station_full_entries = 16,
-    reservation_station_partial_entries = 8,
+    reservation_station_entries_per_type = 16,
```
Comment (hngenc) [inline src/main/scala/gemmini/Configs.scala:54]: Can we get separate lengths for each type (_execute_, _load_, and _store_)?

In general, we need a lot more _execute_ entries than we do for _load_ and _store_ entries.

File: src/main/scala/gemmini/GemminiConfigs.scala [hunk 3]
```diff
@@ -187,9 +186,11 @@ case class GemminiArrayConfig[T <: Data : Arithmetic, U <: Data, V <: Data](
   //==========================================================================
   // cisc-gemmini miscellaneous constants (some redundant with above)
   //==========================================================================
-  val rob_entries      = reservation_station_full_entries + reservation_station_partial_entries
+  val rob_entries      = pow(2, log2Up(reservation_station_entries_per_type)).toInt * 3
```
Comment (hngenc) [inline src/main/scala/gemmini/GemminiConfigs.scala:189]: Why not just make this `reservation_station_entries_per_type * 3`?

File: src/main/scala/gemmini/ReservationStation.scala [hunk 4]
```diff
@@ -54,7 +54,7 @@ class ReservationStation[T <: Data : Arithmetic, U <: Data, V <: Data](config: G
   })
 
   // TODO make this a ChiselEnum
-  val ldq :: stq :: exq :: Nil = Enum(3)
+  val ldq :: exq :: stq :: Nil = Enum(3)
```
Comment (hngenc) [inline src/main/scala/gemmini/ReservationStation.scala:57]: Why did you change the order?

I'm fine with the new order; I'm just curious

File: src/main/scala/gemmini/ReservationStation.scala [hunk 5]
```diff
@@ -281,91 +269,77 @@ class ReservationStation[T <: Data : Arithmetic, U <: Data, V <: Data](config: G
...
-    wars_probe := Cat(wars.reverse)
-    older_in_same_q_probe := Cat(older_in_same_q.reverse)
-    is_st_and_must_wait_for_prior_ex_config_probe := Cat(is_st_and_must_wait_for_prior_ex_config.reverse)
-    is_ex_config_and_must_wait_for_prior_st_probe := Cat(is_ex_config_and_must_wait_for_prior_st.reverse)
+    assert(ldq < exq && exq < stq)
+
+    val not_config = !new_entry.is_config
+    when (is_load) {
+      // war (after ex/st) | waw (after ex)
+      val ld_deps = VecInit(entries_ld.map { e => e.valid && !e.bits.issued }) // same q - can this be ignored?
```
Comment (hngenc) [inline src/main/scala/gemmini/ReservationStation.scala:277]: We should *not* ignore "same-q" checks, so you did the right thing here

File: src/main/scala/gemmini/ReservationStation.scala [hunk 6]
```diff
@@ -281,91 +269,77 @@ class ReservationStation[T <: Data : Arithmetic, U <: Data, V <: Data](config: G
...
+
+      val ex_deps = VecInit(entries_ex.map { e => e.valid && !new_entry.is_config && (
+        (new_entry.opa.bits.overlaps(e.bits.opa.bits) && e.bits.opa.valid) || // waw if preload, war if compute
+        (new_entry.opa.bits.overlaps(e.bits.opb.bits) && e.bits.opb.valid))}) // war
+
+      val st_deps = VecInit(entries_st.map { e => e.valid && e.bits.opa.valid && not_config &&
+        new_entry.opa.bits.overlaps(e.bits.opa.bits)})  // war
+
+      new_entry.deps := ld_deps ++ ex_deps ++ st_deps // TODO: doesn't work when entries per type isnt power of 2
```
Comment (hngenc) [inline src/main/scala/gemmini/ReservationStation.scala:286]: I'm perfectly fine with restricting entries-per-type to be a power of 2.

But you need to add something like an `assert(isPow2(ld_deps.size))` if you're making that assumption

File: src/main/scala/gemmini/ReservationStation.scala [hunk 7]
```diff
@@ -281,91 +269,77 @@ class ReservationStation[T <: Data : Arithmetic, U <: Data, V <: Data](config: G
...
-      full_entries(full_alloc_id).bits := new_entry
-      new_full_allocs(full_alloc_id) := true.B
-    }
+    Seq(
+      (ldq, entries_ld, new_allocs_oh_ld),
+      (exq, entries_ex, new_allocs_oh_ex),
+      (stq, entries_st, new_allocs_oh_st))
+      .foreach { case (q, entries_type, new_allocs_type) =>
+        when (new_entry.q === q) {
+          val is_full = PopCount(Seq(dst.valid, op1.valid, op2.valid)) > 1.U
```
Comment (hngenc) [inline src/main/scala/gemmini/ReservationStation.scala:323]: We don't need an `is_full` check anymore.

But just to be safe, let's keep the `is_full` variable and add this assert right underneath it:

```
if (q != exq) { assert(!is_full) }
```

File: src/main/scala/gemmini/ReservationStation.scala [hunk 8]
```diff
@@ -281,91 +269,77 @@ class ReservationStation[T <: Data : Arithmetic, U <: Data, V <: Data](config: G
...
+          val alloc_id = MuxCase((reservation_station_entries_per_type-1).U, entries_type.zipWithIndex.map { case (e, i) => !e.valid -> i.U })
+
+          when (!entries_type(alloc_id).valid) {
+            io.alloc.ready := true.B
+            entries_type(alloc_id).valid := true.B
+            entries_type(alloc_id).bits := new_entry
+            when (!is_full) { // TODO: can this be removed?
+              entries_type(alloc_id).bits.opb.valid := false.B
+              entries_type(alloc_id).bits.opb.bits := DontCare
+            }
```
Comment (hngenc) [inline src/main/scala/gemmini/ReservationStation.scala:338]: Yes, please remove this if-statement completely.

`opb.valid` is already set above.

File: src/main/scala/gemmini/ReservationStation.scala [hunk 9]
```diff
@@ -389,31 +363,38 @@ class ReservationStation[T <: Data : Arithmetic, U <: Data, V <: Data](config: G
...
-  Seq((ldq, io.issue.ld), (stq, io.issue.st), (exq, io.issue.ex)).foreach { case (q, io) =>
-    val issue_valids = entries.map(e => e.valid && e.bits.ready() && !e.bits.issued && e.bits.q === q)
+  Seq((ldq, io.issue.ld, entries_ld), (exq, io.issue.ex, entries_ex), (stq, io.issue.st, entries_st))
+    .foreach { case (q, io, entries_type) =>
+
+    val issue_valids = entries_type.map(e => e.valid && e.bits.ready() && !e.bits.issued)
     val issue_sel = PriorityEncoderOH(issue_valids)
     val issue_id = OHToUInt(issue_sel)
-    val issue_entry = Mux1H(issue_sel, entries)
+    val global_issue_id = Cat(q.asUInt, issue_id)
```
Comment (hngenc) [inline src/main/scala/gemmini/ReservationStation.scala:372]: Does this assume that the queue-lengths are powers-of-2?

If so, can you add another assert here? Something like this:

```
assert(isPow2(issue_id.getWidth))
```

**Review by hngenc — CHANGES_REQUESTED** [review]

File: src/main/scala/gemmini/ReservationStation.scala (hunk 4 above)
Comment (RCD-Y) [reply src/main/scala/gemmini/ReservationStation.scala:57]: Initially I concatenated the separated lists this way without paying attention the enum order, then the issue id's started to depend on the enum order as well. Between reconcatenating all the lists and changing the enum I chose the latter. As for the initial decision, I thought it made a little more sense because this is usually the workflow (load, use it, then store the result).

Comment (hngenc) [reply src/main/scala/gemmini/ReservationStation.scala:57]: Sounds good; just add a comment saying that the later code assumes this ordering

File: src/main/scala/gemmini/Configs.scala (hunk 2 above)
Comment (RCD-Y) [reply src/main/scala/gemmini/Configs.scala]: Done

File: src/main/scala/gemmini/GemminiConfigs.scala (hunk 3 above)
Comment (RCD-Y) [reply src/main/scala/gemmini/GemminiConfigs.scala]: Is now the sum of the number of entries for the different types

File: src/main/scala/gemmini/ReservationStation.scala (hunk 6 above)
Comment (RCD-Y) [reply src/main/scala/gemmini/ReservationStation.scala]: Addressed at the start of the reservation station; each entry count is tested

File: src/main/scala/gemmini/ReservationStation.scala (hunk 7 above)
Comment (RCD-Y) [reply src/main/scala/gemmini/ReservationStation.scala:320]: Added

File: src/main/scala/gemmini/ReservationStation.scala (hunk 9 above)
Comment (RCD-Y) [reply src/main/scala/gemmini/ReservationStation.scala]: Addressed earlier

File: src/main/scala/gemmini/ReservationStation.scala (hunk 8 above)
Comment (RCD-Y) [reply src/main/scala/gemmini/ReservationStation.scala]: Removed

**Review by hngenc — APPROVED** [review]
LGTM. Great job, @RCD-Y!
